presto-cache/src/main/java/com/facebook/presto/cache/filemerge/FileMergeCacheManager.java

```diff
@@ -161,17 +166,43 @@ public boolean get(FileReadRequest request, byte[] buffer, int offset)
             stats.incrementCacheMiss();
         }
 
-        return result;
+        if (!result && hiveFileContext.getQuota().isPresent() && getCacheSize(hiveFileContext.getTable()) + request.getLength() > hiveFileContext.getQuota().getAsLong()) {
+            // it is not cached and we don't have enough cache quota to store
+            return Optional.empty();
+        }
+
+        try {
+            // hint the cache
+            cache.get(request.getPath(), hiveFileContext::getTable);
+        }
+        catch (ExecutionException e) {
+            // ignore
+        }
+        return Optional.of(result);
+    }
+
+    private long getCacheSize(String table)
+    {
+        long sum = 0;
+        for (Path path : tableFiles.get(table)) {
+            for (Range<Long> range : persistedRanges.get(path).getRange().asDescendingMapOfRanges().keySet()) {
```

Comment:
You need to grab the read lock for range `cacheRange.getLock().readLock()`. `CacheRange` is not thread safe without lock.